### PR TITLE
Add support for the session GUID in the UI

### PR DESCRIPTION
**Pull request by OJ**
## Overview

Associated Payloads PR: https://github.com/rapid7/metasploit-payloads/pull/203
Associated Mettle PR: https://github.com/rapid7/mettle/pull/82

MSF doesn't currently attempt to track sessions in any way, other than through the UUID which identifies various things with respect to the arch/platform. This isn't enough. We don't have the ability to know if a session has come back to live, nor do we have a way of identifying a single active session across multiple transports.

The aim of the PR is to start solving this problem. It adds a per-session identifier (a GUID) to Meterpreter sessions.

* For staged payloads, the stage is generated with the session GUID baked in.
* For stageless payloads, the generated payload contains a blank guid. Once the new stageless session comes in, a GUID is generated on the fly and passed to the Meterpreter instance.

The aim down the track will be to start using this GUID as a means for accounting. From here we'll be able to do some more interesting things including:

* Tracking of sessions that come alive again after being put to sleep.
* Having TLV packets make their way to MSF through multiple transports.
* Allow us to start separating transport from session.
* Allow Meterpreter to act as a pivot point by passing on TLV packets from other sessions without the need to wrap them in channels.

For starters, there's a new Meterpreter command called `guid` which shows the session guid. This isn't really practically useful for now, but it's there to allow us to validate the guid once the sessions are in place. This will probably be removed down the track.

This PR is just the beginning. Much more to come.

## Verification

- [x] Create listeners and payloads for ALL THE METERPRETERS on all platforms, both staged and stageless.
- [x] Create payloads that match ^^.
- [x] Fire off each of these payloads and make sure they connect properly.
- [x] For each new session, make sure that the `guid` command gives a guid that is unique to the session.
- [x] Fire the same stageless payload and make sure the guid is different from the same payload.
- [x] Put some transports to sleep, or switch transports, or kill the connections and let them reestablish and make sure that the guids are the same as they were once the session comes back in.

### CODE REVIEW

**Comment by busterb** [comment]
Question: GUID is basically a unique session identifier? If so, would it be easier to call it SID or SESSIONID?

**Comment by OJ** [comment]
The main reason for it was [this](https://github.com/rapid7/metasploit-framework/blob/d8f59a9d622ff37db9392bbd104d7456f2ff8c7e/lib/msf/core/session.rb#L340):
``` ruby
  #
  # The session unique identifier.
  #
  attr_accessor :sid
```
I wanted to use `sid` or `session_id` but it would have conflated the meaning with the existing one, so I explicitly used `session_guid` and made that consistent across everything.

**Comment by OJ** [comment]
BTW, the build here is failing because of payload sizes. When we're ready to merge and we have the Payloads PR landed, I'll update the payload sizes then.

**Comment by bwatters-r7** [comment]
I'm in the process of updating the test suite, but this sort of screams automated testing to a degree.  Checking GUIDs will mean manually parsing the resultant session data, but that's relatively minor compared to payload generation and running.  If this is not in a terrible hurry, I'll try and have something for Windows set up by Friday.

**Comment by bwatters-r7** [comment]
Hrm..... I'll also need to allow it to grab payload PRs and install them.... there's a logical path I made sure existed to support that, but I just have not coded it, yet.

**Comment by OJ** [comment]
Added the mettle-related stuffs!

**Comment by busterb** [comment]
Feel free to bump to metasploit-payloads 1.2.33 and metasploit_payloads-mettle 0.1.10

**Comment by OJ** [comment]
Thanks @busterb, gems updated, payload cached sizes updated too. The build _should_ go green now!

**Comment by busterb** [comment]
Awesome! OK Brendan, let er' rip!

File: lib/msf/base/sessions/meterpreter_options.rb [hunk 1]
```diff
@@ -51,6 +51,17 @@ def on_session(session)
     end
 
     if valid
+      # always make sure that the new session has a new guid if it's not already known
+      guid = session.core.get_session_guid
```
Comment (timwr) [inline lib/msf/base/sessions/meterpreter_options.rb]: won't this break compatibility with older (stageless) payloads?

Comment (OJ) [inline lib/msf/base/sessions/meterpreter_options.rb]: Yes.

**Comment by bwatters-r7** [comment]
I do not have cloned target VMs set up yet for the new range, so I'm overloading the number of payloads per target.  That causes some sessions to fail to establish, but since the payloads are duplicated across targets, one or two-off issues are probably acceptable and likely even predictable.  This is just a test of the GUID command.
Windows 7 Targets with reverse (non-http) payloads:

image

**Comment by bwatters-r7** [comment]
# Windows 7 Targets with x86 HTTP-ish payloads:
image

**Comment by bwatters-r7** [comment]
# Windows 7 VMs and Python Payloads on Python27 x86
(Uhhh, I appear to have broken bind payloads again.  I'll look into that later.)
image

**Comment by bwatters-r7** [comment]
Bug Fixed.....
image

**Comment by OJ** [comment]
So does that mean the rc4 and dns payloads are somehow broken?

**Comment by bwatters-r7** [comment]
@OJ Most likely not.  If they were, I would expect them to fail across WIndows 7 versions.  That's most likely due to resource constraints on the VMs because I was overloading the VMs.  I can retest tomorrow now that I cloned out some extra VMs.

**Comment by bwatters-r7** [comment]
@OJ: retested this morning with clones.  All green.  Going to keep going today.

**Comment by bwatters-r7** [comment]
# Python payloads on Win 7 and Python 2.7x64:
image

**Comment by bwatters-r7** [comment]
Found a bug.....
So if you use a python payload on python 3 (in this case python 3.4.3 (x86 or x64) the connection starts out OK but at some point fails when a call is made to ``decode`` on the string class which is deprecated in python 3.

It will run fine if you just ryun it; it is part of the handshake (likely the guid exchange) that causes the crash.

msfconsole output:
```
[ruby-2.4.1@metasploit-framework]((HEAD detached at upstream/pr/8523)) tmoose@ubuntu:~/rapid7/metasploit-framework$ ./msfconsole
       =[ metasploit v4.14.25-dev-c4288fb                 ]
msf > use exploit/multi/handler 
msf exploit(handler) > set payload python/meterpreter/reverse_tcp
payload => python/meterpreter/reverse_tcp
msf exploit(handler) > set lport 4568
lport => 4568
msf exploit(handler) > set lhost 192.168.135.111
lhost => 192.168.135.111
msf exploit(handler) > run

[*] Started reverse TCP handler on 192.168.135.111:4568 
[*] Starting the payload handler...
[*] Sending stage (40516 bytes) to 192.168.134.148
[*] Meterpreter session 1 opened (192.168.135.111:4568 -> 192.168.134.148:49241) at 2017-06-15 15:59:45 -0500
[*] 192.168.134.148 - Meterpreter session 1 closed.  Reason: Died

[-] Invalid session identifier: 1
msf exploit(handler) > run

[*] Started reverse TCP handler on 192.168.135.111:4568 
[*] Starting the payload handler...
[-] Meterpreter session 1 is not valid and will be closed

```
output on WIndows machine:
image

**Comment by OJ** [comment]
Thanks @bwatters-r7 I'll take a look!

**Comment by OJ** [comment]
What's odd about this is that I'm not doing anything different to what we've already done with the UUID in the past. I wonder of this breaks on master as well ...

**Comment by OJ** [comment]
Oh I think I see it.

**Comment by OJ** [comment]
I suck at git. Please ignore this noise while I fix it.

**Comment by OJ** [comment]
OK @bwatters-r7 I have fixed my Python3 faux pas. Sorry about that. Should be good in both Python3 and Python2. Thanks very much for catching that! I've updated the payload cached sizes too.

**Comment by bwatters-r7** [comment]
That did it.....
# Windows 7 Python 3 x64
image

**Comment by bwatters-r7** [comment]
Now that the range is scaled up, the reports will not fit on the page anymore.
All possible combinations of the following OS/Payload/Interpreter passed:

# Commit:
9ce0bb9345800eb5e139deeae90ba1a86141deae

# Operating systems:
Windows 7x86 SP0
Windows 7x86 SP1
Windows 7x64 SP0
Windows 7x64 SP1

# Payloads:
## *.exe
windows/meterpreter/bind_tcp
windows/meterpreter/bind_tcp_rc4
windows/meterpreter/bind_tcp_uuid
windows/meterpreter/bind_tcp
windows/x64/meterpreter/bind_tcp
windows/meterpreter_bind_tcp
windows/x64/meterpreter_bind_tcp
windows/meterpreter/reverse_tcp
windows/meterpreter/reverse_tcp_rc4
windows/meterpreter/reverse_tcp_rc4_dns
windows/meterpreter/reverse_tcp_dns
windows/meterpreter/reverse_tcp_uuid
windows/meterpreter_reverse_tcp
windows/meterpreter/reverse_http
windows/meterpreter/reverse_https
windows/meterpreter_reverse_http
windows/meterpreter_reverse_https
windows/meterpreter/reverse_winhttp
windows/meterpreter/reverse_winhttps
windows/x64/meterpreter/reverse_http
windows/x64/meterpreter/reverse_https
windows/x64/meterpreter/reverse_tcp
windows/x64/meterpreter/reverse_tcp_uuid
windows/x64/meterpreter/reverse_winhttp
windows/x64/meterpreter/reverse_winhttps
windows/x64/meterpreter_reverse_tcp

## *.py
(Tested on python 2.7 x86 | python 2.7 x64 | python 3.4 x86 | python 3.4 x64)
python/meterpreter/bind_tcp
python/meterpreter/bind_tcp_uuid
python/meterpreter/reverse_http
python/meterpreter/reverse_tcp
python/meterpreter/reverse_tcp_ssl
python/meterpreter/reverse_tcp_uuid
python/meterpreter_bind_tcp
python/meterpreter_reverse_http
python/meterpreter_reverse_https
python/meterpreter_reverse_tcp

## *.jar
(Tested on Java x86 | Java x64)
java/meterpreter/bind_tcp
java/meterpreter/reverse_http
java/meterpreter/reverse_https
java/meterpreter/reverse_tcp

**Comment by busterb** [comment]
Landed, tests looking good, the range is still in progress, but we're pretty confident now. Thanks!

**Comment by busterb** [comment]
## Release Notes

This change adds foundational support for tracking a single active session across multiple transports changes, connections, and tunneling mechanisms. It is the first step in disassociating connections from the sessions.

**Pull request merged**